api/envoy/extensions/filters/http/json_to_metadata/v3/json_to_metadata.proto

```diff
@@ -0,0 +1,121 @@
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Json-To-Metadata Filter]
+//
+// The configuration for transforming json body into metadata. This is useful
+// for matching load balancer subsets, logging, etc.
+//
+// Json to Metadata :ref:`configuration overview <config_http_filters_json_to_metadata>`.
+// [#extension: envoy.filters.http.json_to_metadata]
+
+message JsonToMetadata {
+  enum ValueType {
+    // The value is a serialized `protobuf.Value
+    // <https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/struct.proto#L62>`_.
+    PROTOBUF_VALUE = 0;
+
+    STRING = 1;
+
+    NUMBER = 2;
+  }
+
+  // [#next-free-field: 6]
+  message KeyValuePair {
+    // The namespace — if this is empty, the filter's namespace will be used.
+    string metadata_namespace = 1;
+
+    // The key to use within the namespace.
+    string key = 2 [(validate.rules).string = {min_len: 1}];
+
+    oneof value_type {
+      // The value to pair with the given key.
+      //
+      // When used for on_present case, if value is non-empty it'll be used instead
+      // of the attribute value. If both are empty, the attribute value is used as-is.
+      //
+      // When used for on_missing/on_error case, a non-empty value
+      // must be provided.
+      //
+      // It ignores ValueType, i.e., not type casting.
+      google.protobuf.Value value = 3;
+    }
+
+    // The value's type — defaults to protobuf.Value.
+    ValueType type = 4 [(validate.rules).enum = {defined_only: true}];
+
+    // False if we want to overwrite the existing metadata value. Default to false.
+    bool preserve_existing_metadata_value = 5;
+  }
+
+  message Selector {
+    // TODO(kuochunghsu): Explore matchers for array handling.
+    oneof selector {
+      // key to match
+      string key = 1 [(validate.rules).string = {min_len: 1}];
+    }
+  }
+
+  // A Rule defines what metadata to apply when a key-value is present, missing in the json
+  // or fail to parse the payload.
+  message Rule {
+    // Specifies that a match will be performed on the value of a property.
+    // Here's an example to match on 1 in {"foo": {"bar": 1}, "bar": 2}
+    //
+    // selectors:
+    // - key: foo
+    // - key: bar
+    repeated Selector selectors = 1 [(validate.rules).repeated = {min_items: 1}];
+
+    // If the attribute is present, apply this metadata KeyValuePair.
+    KeyValuePair on_present = 2;
+
+    // If the attribute is missing, apply this metadata KeyValuePair.
+    //
+    // The value in the KeyValuePair must be set.
+    KeyValuePair on_missing = 3;
+
+    // If the body is too large or fail to parse, apply this metadata KeyValuePair.
+    //
+    // The value in the KeyValuePair must be set.
+    KeyValuePair on_error = 4;
+  }
+
+  // the list of rules to apply to requests
+  repeated Rule request_rules = 1 [(validate.rules).repeated = {min_items: 1}];
```

Comment:
Instead of building your own matching API, have you considered using the [xDS generic matching API](https://github.com/cncf/xds/blob/main/xds/type/matcher/v3/matcher.proto)?  See the original design [here](https://docs.google.com/document/d/1G4g-6q0IArz_ERgqixzZCM0-wbexFYuBUvT9hkV7QRU/edit).

Reply:
We haven't walked through various matchers but seek for MVP while we implemented this. I just did a quick check the xds matching API that current utilities are around http requests? Json matchers might have very different natures and we are happy on current design as it fits the needs. Does it make sense?